### PR TITLE
Bryce davidson/vir 1986 use subtraction faking

### PR DESCRIPTION
**Pull request by Bryce-Davidson**
## Pre-Review Checklist
* [x] I have considered backwards and forwards compatibility.
* [x] All changes are tested.
* [x] All touched code documentation is updated.
* [x] Deepsource issues have been reviewed and addressed.
* [x] Comments and `print` statements have been removed.

### CODE REVIEW

File: tests/subtractions/test_db.py [hunk 1]
```diff
@@ -1,61 +1,83 @@
...
-    await mongo.subtraction.insert_many(
-        [{"_id": "foo", "name": "Foo"}, {"_id": "bar", "name": "Bar"}], session=None
-    )
+    ]
+
+    for document in documents:
+        result = await apply_transforms(document, [AttachSubtractionsTransform(mongo)])
 
-    result = await apply_transforms(documents, [AttachSubtractionsTransform(mongo)])
+        assert result == snapshot
```
Comment (igboyes) [inline tests/subtractions/test_db.py:36]: The way this was before was faster.

**Review by igboyes — CHANGES_REQUESTED** [review]

File: tests/subtractions/test_db.py (hunk 1 above)
Comment (Bryce-Davidson) [reply tests/subtractions/test_db.py:36]: Not sure how to pass the created subtraction Id to the parameterized test. I guess I should just write a separate function to pass to the `pytest.mark.parametrize` so the tests can be parallelized?

Comment (Bryce-Davidson) [reply tests/subtractions/test_db.py:36]: Do you want me to break this up into multiple tests? There is no straight forward way to get code execution within the `pytest.mark.parametrize` that works nicely with the new fake API.

Comment (igboyes) [reply tests/subtractions/test_db.py:36]: This test isn't parametrized is it?

I am just saying the last few lines can look like this instead:
```py
assert await apply_transforms(document, [AttachSubtractionsTransform(mongo)]) == snapshot
```

The `apply_transforms` function can take either single document or list of documents as its first argument.

File: tests/subtractions/test_api.py [hunk 2]
```diff
@@ -144,361 +108,555 @@ async def test_get_from_job(fake: DataFaker, spawn_job_client, snapshot_recent):
...
+        test_dir: Path = tmp_path / "files"
+        test_dir.mkdir(exist_ok=True)
+
+        file_directory: Path = test_dir
+        subtraction_directory: Path = tmp_path / "subtractions"
+
+        return self.Context(
+            file_directory=file_directory,
+            subtraction_directory=subtraction_directory,
         )
```
Comment (igboyes) [inline tests/subtractions/test_api.py:213]: This is only used in `test_create` and doesn't need to be a fixture.

---
In fact, I don't think it needs to exist at all.

When you use `spawn_job_client` an application server is created with it own temporary data directory (same as the `--data-path` configuration value used when starting the server.

The `data_path` is accessible using the eponymous fixture. This means the following can be simplified:
```py
# Initial
assert os.listdir(
    ctx.subtraction_directory / subtraction.id,
) == [
    self.VALID_SUBTRACTION_FILE_NAME,
]

# With ``data_path`` fixture.
assert os.listdir(data_path / "subtractions" / subtraction.id) == [
    self.VALID_SUBTRACTION_FILE_NAME,
]
```

Let me know if I am missing some nuance in this case.

File: tests/subtractions/test_api.py [hunk 3]
```diff
@@ -144,361 +108,555 @@ async def test_get_from_job(fake: DataFaker, spawn_job_client, snapshot_recent):
...
+        - test_subtraction_file_invalid_name_error: Ensures proper handling of invalid file names.
+        - test_subtraction_file_name_conflict_error: Verifies handling of file name conflicts.
+    """
+
+    VALID_SUBTRACTION_FILE_NAME = "subtraction.1.bt2"
+    INVALID_SUBTRACTION_FILE_NAME = "reference.1.bt2"
+
+    @dataclass
+    class Context:
+        file_directory: Path
```
Comment (igboyes) [inline tests/subtractions/test_api.py:196]: No used as far as I can tell.

File: tests/subtractions/test_api.py [hunk 4]
```diff
@@ -144,361 +108,555 @@ async def test_get_from_job(fake: DataFaker, spawn_job_client, snapshot_recent):
...
+        user = await fake.users.create()
+        upload = await fake.uploads.create(
+            user=user,
+            upload_type=UploadType.subtraction,
+        )
+        subtraction = await fake.subtractions.create(user=user, upload=upload)
+
+        Path.unlink(
+            data_path / "subtractions" / subtraction.id / "subtraction.1.bt2",
+        )
```
Comment (igboyes) [inline tests/subtractions/test_api.py:632]:
```suggestion
        (data_path / "subtractions" / subtraction.id / "subtraction.1.bt2").unlink()
```

File: tests/subtractions/test_db.py [hunk 5]
```diff
@@ -2,60 +2,87 @@
...
-    [
-        {"id": "sub_1", "subtractions": ["foo", "bar"]},
-        [
-            {"id": "sub_1", "subtractions": ["foo", "bar"]},
-            {"id": "sub_2", "subtractions": ["foo"]},
-            {"id": "sub_3", "subtractions": []},
-        ],
-    ],
+    "case",
+    ["single", "multiple"],
```
Comment (igboyes) [inline tests/subtractions/test_db.py:15]: I see what you were saying about this test.

I think the most elegant solution is to convert to a class-based test with `test_single` and `test_multi` methods.

**Review by igboyes — CHANGES_REQUESTED** [review]

File: tests/subtractions/test_db.py (hunk 5 above)
Comment (Bryce-Davidson) [reply tests/subtractions/test_db.py:15]: With the parameterization it has now it runs in parallel when using`xdist`

File: tests/subtractions/test_api.py (hunk 2 above)
Comment (Bryce-Davidson) [reply tests/subtractions/test_api.py:213]: Yes, sorry. An artifact of when I thought I needed it in all tests. removing.

File: tests/subtractions/test_db.py (hunk 5 above)
Comment (igboyes) [reply tests/subtractions/test_db.py:15]: It should either way.

**Review by igboyes — APPROVED** [review]

**Pull request merged**